### PR TITLE
Eliminate intermediate string and byte[] allocation in XML marshallers

### PR DESCRIPTION
**Pull request by muhammad-othman**
## Description

The previous XML marshalling path performed two unnecessary allocations per request:

```
XmlWriter → XMLEncodedStringWriter (StringWriter) → string → Encoding.UTF8.GetBytes(string) → byte[] → request.Content
```

The new path writes UTF-8 bytes directly into the pooled buffer:

```
XmlWriter → XMLEncodedBufferTextWriter → ArrayPoolBufferWriter<byte> (inside PooledContentStream)
```

This eliminates both the intermediate UTF-16 string and the final `byte[]` copy, building on `PooledContentStream` introduced in #4409.

To achieve this I've added `XMLEncodedBufferTextWriter` which is a `TextWriter` that:
1. Intercepts characters from `XmlWriter` and applies the same AWS-specific escaping as `XMLEncodedStringWriter` (`\n` → `&#xA;`, `\^E` → `&#x85;`, `\u2028` → `&#x2028;`)
2. UTF-8 encodes characters directly into the `ArrayPoolBufferWriter<byte>` via `System.Text.Encoder`
3. On .NET 8+, uses the Span-based `Encoder.Convert` overload to write directly into the buffer with zero intermediate copy
4. Batch-encodes runs of non-special characters in a single `Encoder.Convert` call for throughput
5. Correctly handles surrogate pairs (emoji, etc.) by deferring flush until both halves arrive

## Motivation and Context
`DOTNET-8683`

## Testing
- Added `XMLEncodedBufferTextWriterTests.cs`
- Benchmarked with Route53 `ChangeResourceRecordSets` and S3 `PutLifecycleConfiguration` across varying payload sizes.

**Allocation savings scale with payload size (up to 66% reduction for large payloads), with no latency regression.**

  Here are the results before this change:

| Method                   | RecordCount | Mean      | Error     | StdDev   | Median    | P50       | P90      | Allocated |
|------------------------- |------------ |----------:|----------:|---------:|----------:|----------:|---------:|----------:|
| ChangeResourceRecordSets | 10          |  81.76 ms |  4.401 ms | 12.56 ms |  78.63 ms |  78.63 ms | 100.3 ms | 132.71 KB |
| ChangeResourceRecordSets | 100         | 180.41 ms |  8.923 ms | 26.03 ms | 174.63 ms | 174.63 ms | 222.2 ms | 317.41 KB |
| ChangeResourceRecordSets | 250         | 324.51 ms |  9.309 ms | 27.15 ms | 322.97 ms | 322.97 ms | 358.7 ms | 508.22 KB |
| ChangeResourceRecordSets | 500         | 455.74 ms | 12.993 ms | 37.28 ms | 444.53 ms | 444.53 ms | 513.1 ms | 861.66 KB |

| Method                    | RuleCount | Mean     | Error    | StdDev   | Median   | P50      | P90      | Allocated  |
|-------------------------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|-----------:|
| PutLifecycleConfiguration | 10        | 211.9 ms | 15.69 ms | 46.27 ms | 195.9 ms | 195.9 ms | 284.0 ms |  119.81 KB |
| PutLifecycleConfiguration | 100       | 206.8 ms |  6.88 ms | 19.85 ms | 203.1 ms | 203.1 ms | 234.7 ms |  296.01 KB |
| PutLifecycleConfiguration | 500       | 263.7 ms |  9.37 ms | 26.58 ms | 262.6 ms | 262.6 ms | 302.1 ms | 1013.27 KB |
| PutLifecycleConfiguration | 1000      | 359.5 ms | 17.77 ms | 52.12 ms | 358.7 ms | 358.7 ms | 429.8 ms | 1933.55 KB |

And after this change:

| Method                   | RecordCount | Mean      | Error     | StdDev   | P50       | P90      | Allocated |
|------------------------- |------------ |----------:|----------:|---------:|----------:|---------:|----------:|
| ChangeResourceRecordSets | 10          |  88.43 ms |  6.296 ms | 18.27 ms |  84.43 ms | 115.7 ms | 131.39 KB |
| ChangeResourceRecordSets | 100         | 175.71 ms |  8.272 ms | 24.13 ms | 170.43 ms | 216.7 ms | 234.25 KB |
| ChangeResourceRecordSets | 250         | 250.63 ms |  6.354 ms | 18.13 ms | 246.97 ms | 278.0 ms | 352.68 KB |
| ChangeResourceRecordSets | 500         | 466.21 ms | 13.363 ms | 38.77 ms | 461.38 ms | 528.7 ms | 614.77 KB |

| Method                    | RuleCount | Mean     | Error    | StdDev   | P50      | P90      | Allocated |
|-------------------------- |---------- |---------:|---------:|---------:|---------:|---------:|----------:|
| PutLifecycleConfiguration | 10        | 181.2 ms |  3.85 ms | 11.11 ms | 179.5 ms | 198.7 ms |  124.7 KB |
| PutLifecycleConfiguration | 100       | 199.8 ms |  5.43 ms | 15.32 ms | 197.7 ms | 219.9 ms | 134.37 KB |
| PutLifecycleConfiguration | 500       | 266.5 ms |  9.35 ms | 27.11 ms | 264.4 ms | 301.7 ms | 375.25 KB |
| PutLifecycleConfiguration | 1000      | 321.6 ms | 14.66 ms | 41.58 ms | 323.2 ms | 370.2 ms | 650.49 KB |

### Dry-runs

- **DotNet Dry-run ID:** 094d25a9-a7ac-42b3-8bfd-e65192406c72
  - [ ] Pending
  - [x] Completed successfully
  - [ ] Failed
- **PowerShell Dry-run ID:** e61a854a-1f85-4f57-8c46-941d68337f9d
  - [ ] Pending
  - [x] Completed successfully
  - [ ] Failed

## Breaking Changes Assessment

1. Identify all breaking changes including the following details:
    * What functionality was changed?
    * How will this impact customers?
    * Why does this need to be a breaking change and what are the most notable non-breaking alternatives?
    * Are best practices being followed?
    * How have you tested this breaking change?
2. Has a senior/+ engineer been assigned to review this PR?

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/test/Services/S3/UnitTests/Custom/LifecycleConfigurationTests.cs [hunk 1]
```diff
@@ -208,7 +208,12 @@ public void TestPutLifecycleConfiguration()
                 },
             };
             IRequest iRequest = PutLifecycleConfigurationRequestMarshaller.Instance.Marshall(request);
-            string content = Encoding.UTF8.GetString(iRequest.Content);
+            var contentBytes = iRequest.Content
+#if !NETFRAMEWORK
+                ?? ((PooledContentStream)iRequest.ContentStream).Content.ToArray()
+#endif
+                ;
+            string content = Encoding.UTF8.GetString(contentBytes);
```
[the remark this answers is not in the document]
Comment (muhammad-othman) [reply sdk/test/Services/S3/UnitTests/Custom/LifecycleConfigurationTests.cs]: Added `using` to the request.

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlRequestMarshaller.tt [hunk 2]
```diff
@@ -332,13 +339,15 @@ WriteXmlAttributeString(level + 1, member, variableName, isPayload: true, operat
 			PostMarshallCustomization(request, publicRequest);
 			try 
 			{
+#if NETFRAMEWORK
```
Comment (normj) [inline generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlRequestMarshaller.tt]: nit: Nice to have for future readers is a suffix comment explaining why the non .NET Framework targets are covered. Other wise readers reading this code have to tie together the logic made much further away.

```
#if NETFRAMEWORK // For non .NET Framework targets the request payload is stored in the ContentStream via the PooledContentStream.
...
```

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        public override void Write(string value)
+        {
+            if (value == null) return;
+
+            WriteSpan(value.AsSpan());
+        }
+
+        // Scans for special characters, batch-encodes clean runs between them in a single
```
Comment (normj) [inline sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: I'm missing the context to understand the encoding here because I don't understand what a "clean run" is. Can the comments provide more context.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        // Batch-encodes a clean run of characters (no special chars) into the pooled buffer.
+        private void EncodeSpan(ReadOnlySpan<char> chars)
+        {
+#if NET8_0_OR_GREATER
+            // Span-based overload writes directly into the pooled buffer — zero intermediate copies.
+            while (chars.Length > 0)
+            {
+                var dest = _writer.GetSpan(chars.Length * 3); // worst case: 3 UTF-8 bytes per char
```
Comment (normj) [inline sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: Isn't 4 bytes per UTF-8 character the worst case?

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs (hunk 3 above)
Comment (muhammad-othman) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: I meant runs without any characters that needs escaping, will update the comment to make it more readable.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs (hunk 4 above)
Comment (muhammad-othman) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: Not really, 4 bytes will be encoded as two char values, these are the surrogate pair that I keep mentioning around the flush. 
This matches the .NET runtime's own constant [MaxUtf8BytesPerChar = 3](https://github.com/dotnet/runtime/blob/35a593a24724ce4b73f56963ced82a4ecb66cb5b/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs#L57), which has the this comment `Supplementary code points are expanded to UTF-8 from UTF-16 at a 4:2 ratio, so 3:1 is still the correct value for maximum expansion.`

File: generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlRequestMarshaller.tt (hunk 2 above)
Comment (muhammad-othman) [reply generator/ServiceClientGeneratorLib/Generators/Marshallers/RestXmlRequestMarshaller.tt]: Updated.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs (hunk 3 above)
Comment (muhammad-othman) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: Updated the comment with more context.

File: sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs (hunk 4 above)
Comment (muhammad-othman) [reply sdk/src/Core/Amazon.Runtime/Internal/Util/XMLEncodedBufferTextWriter.cs]: Added more context to the comment.

**Review by peterrsongg — APPROVED** [review]

**Comment by peterrsongg** [comment]
ran the tests on my own machine, with a few more edge cases I added as well and they passed. so approving.

**Review by normj — APPROVED** [review]
